### PR TITLE
tags/findSeries - add lastts-json format

### LINKED ISSUE #1570: Support efficient "lastUpdate" endpoint
We have a use case (detecting absence of metrics) where we want an efficient method of querying for a the lastUpdate value for series that match an expression. Specifically, we don't care what the last value was, just the most recent timestamp. This means that the query should be very efficient since it never needs to fetch any data. 

Existing endpoint `/tags/findSeries` is close, but don't offer the `lastUpdate` value. 

I am willing to implement the endpoint, but submitting the issue to kickoff design discussion / feedback.

### PR DESCRIPTION
**Pull request by shanson7**
This is to support efficient querying for missing data. 

Fix #1570

### CODE REVIEW

**Comment by Dieterbe** [comment]
Update http docs please

**Comment by shanson7** [comment]
I just added a commit that will return partial results if the query is too large. I think this is ok for the `findSeries` endpoint (the `render` behavior will still error out). Do you think that I should add an explicit `limit` to the request (and error if it's too large)? Or return some metadata that indicates the results are truncated?

File: api/graphite.go [hunk 1]
```diff
@@ -38,6 +38,10 @@ var MissingOrgHeaderErr = errors.New("orgId not set in headers")
 var MissingQueryErr = errors.New("missing query param")
 var InvalidFormatErr = errors.New("invalid format specified")
 var InvalidTimeRangeErr = errors.New("invalid time range requested")
+var TooManySeriesErr = response.NewError(
+	http.StatusRequestEntityTooLarge,
```
Comment (Dieterbe) [inline api/graphite.go]: i was going to say, this is only for request entities (requests from client, e.g. file uploads), but we do the same for `errMaxPointsPerReq`, so ok...

File: api/models/graphite.go [hunk 2]
```diff
@@ -117,14 +117,25 @@ type GraphiteTagDetailsValueResp struct {
 }
 
 type GraphiteTagFindSeries struct {
-	Expr []string `json:"expr" form:"expr"`
-	From int64    `json:"from" form:"from"`
+	Expr   []string `json:"expr" form:"expr"`
+	From   int64    `json:"from" form:"from"`
+	Format string   `json:"format" form:"format" binding:"Default(seriesjson)"`
```
Comment (Dieterbe) [inline api/models/graphite.go]: 1) should it not be `series-json` ? per the docs? 
2) seems to be missing validation.
this should work:
```suggestion
	Format string   `json:"format" form:"format" binding:"In(,series-json,lastupdate-json);Default(series-json)"`
```

File: docs/http-api.md [hunk 3]
```diff
@@ -66,6 +66,47 @@ json and treejson are the same.
...
+Returns metrics which match tag queries and have received an update since `from`.
+Note: the returned results are not deduplicated and in certain cases it is possible
+that duplicate entries will be returned.
+
+##### Parameters
+
+* expr (required): a list of [tag expressions](#tag-expressions)
+* from: Graphite [from time specification](#fromto) (optional. defaults to now-24hours)
+* format: series-json, lastupdate-json. (defaults to series-json)
+* limit: max number to return, 0 to defer to the instance configured max (default: 0). Note: if limit is 0 or greater than the instance configured max and the result set is greater than the instance configured max, an error is returned. Otherwise, the result set is truncated at the limit
```
Comment (Dieterbe) [inline docs/http-api.md]: "instance configured max" sounds too vague. Let's say `http.max-series-per-req` (note, only for these docs, for the cloud docs above it's good as is, as we don't want to expose those details to our cloud users)

File: api/models/graphite.go (hunk 2 above)
Comment (shanson7) [reply api/models/graphite.go]: 1. Yes, I can change that.
2. I didn't want to add that validation since `format` is currently ignored if specified. Adding the validation means breaking queries that used to work.

File: docs/http-api.md [hunk 4]
```diff
@@ -66,6 +66,47 @@ json and treejson are the same.
...
+
+Returns metrics which match tag queries and have received an update since `from`.
+Note: the returned results are not deduplicated and in certain cases it is possible
+that duplicate entries will be returned.
+
+##### Parameters
+
+* expr (required): a list of [tag expressions](#tag-expressions)
+* from: Graphite [from time specification](#fromto) (optional. defaults to now-24hours)
+* format: series-json, lastupdate-json. (defaults to series-json)
```
Comment (Dieterbe) [inline docs/http-api.md]: opinion:
1) lastUpdate (and lastWrite for that matter) are poorly named variables. ([details](https://github.com/grafana/metrictank/blob/master/docs/metadata.md#lastsave-vs-lastupdate)). This rename would make sense:
  - lastUpdate -> lastTs
  - lastWrite -> lastUpdate
2) "lastUpdate-json" -> not immediately clear what this means exactly.

it seems like you agree with point 1, considering you use "lastTs" in the response data.
I suggest we use format "lastts-json", that way it's consistent with your output and also solves my point 2.
we can use "lastUpdate" internally in the code and internal api's, but everything userfacing can use lastTs/lastts/lastTimestamp.

File: api/models/graphite_gen.go [hunk 5]
```diff
@@ -372,6 +372,356 @@ func (z *GraphiteTagDelSeriesResp) Msgsize() (s int) {
 	return
 }
 
+// DecodeMsg implements msgp.Decodable
```
Comment (Dieterbe) [inline api/models/graphite_gen.go]: why are we generating all the msgp code, tests and benchmarks? we only use json, so let's add an ignore rule for this type.

File: api/graphite.go [hunk 6]
```diff
@@ -1025,14 +1044,34 @@ func (s *Server) graphiteTagFindSeries(ctx *middleware.Context, request models.G
...
+		for _, serie := range series {
+			var lastUpdate int64
+			for _, node := range serie.Series {
+				for _, ndef := range node.Defs {
+					if ndef.LastUpdate > lastUpdate {
+						lastUpdate = ndef.LastUpdate
+					}
+				}
+			}
+			seriesVals = append(seriesVals, models.SeriesLastUpdate{Series: serie.Pattern, Ts: lastUpdate})
```
Comment (Dieterbe) [inline api/graphite.go]: i'm a bit confused here. doesn't this only include the series name (without tags) ? I thought you wanted it broken down by tags.
also, it looks like this doesn't include meta tags.

File: api/graphite.go [hunk 7]
```diff
@@ -1048,11 +1087,19 @@ func (s *Server) clusterFindByTag(ctx context.Context, orgId uint32, expressions
...
+				// Fill in up to maxSeries
+				for _, series := range resp.Metrics[:remainingSpace] {
+					allSeries = append(allSeries, Series{
+						Pattern: series.Path,
+						Node:    r.peer,
+						Series:  []idx.Node{series},
+					})
+				}
+			}
+			return allSeries, TooManySeriesErr
```
Comment (Dieterbe) [inline api/graphite.go]: you want to excute this return even in the softlimit case?

File: api/graphite.go (hunk 6 above)
Comment (shanson7) [reply api/graphite.go]: This returns the same string that is currently returned, which is formatted like `name;tag1=val1;tag2=val2`

File: api/models/graphite.go (hunk 2 above)
Comment (Dieterbe) [reply api/models/graphite.go]: you sure? note that the validation allows an empty string. we use this mechanism elsewhere too.

File: api/graphite.go (hunk 7 above)
Comment (shanson7) [reply api/graphite.go]: Yes. When soft limited, it just returns the results up to the limit

Comment (Dieterbe) [reply api/graphite.go]: but it will also error

File: api/models/graphite.go (hunk 2 above)
Comment (shanson7) [reply api/models/graphite.go]: Empty string would allow it to not exist, right? But what about `format=json`? Wouldn't that error with the validation (but doesn't error currently).

File: api/graphite.go (hunk 7 above)
Comment (shanson7) [reply api/graphite.go]: In the calling function we currently compare the error and ignore it. But I agree that it's cleaner to just not error.

File: api/models/graphite.go (hunk 2 above)
Comment (Dieterbe) [reply api/models/graphite.go]: > Empty string would allow it to not exist, right?

I believe so, but better try it and confirm.

> But what about format=json? Wouldn't that error with the validation (but doesn't error currently).

yes, but no one should be doing that as the format parameter never existed before.
If I'm missing something, you could also just add "json" to the list of accepted values and add it to the switch statement.

**Comment by Dieterbe** [comment]
> Or return some metadata that indicates the results are truncated?

this.
I think any time any of our http endpoints returns data along with a warning, it should be structured as a dictionary, with a warnings field, like so:
```
{
    "series": [
        {
            "lastTs": 1576683990,
            "val": "disk.used;datacenter=dc1;rack=a1;server=web01"
        }
    ],
    "warnings": [
        "result set truncated to 1234 items",
    ]
}

**Comment by shanson7** [comment]
> I think any time any of our http endpoints returns data along with a warning, it should be structured as a dictionary, with a warnings field

That's what I was thinking, but I can't really change the existing (and graphite compatible) `series-json` format to include the warning. I can add the `warning` element to the lastts format though.

**Comment by Dieterbe** [comment]
perhaps it should also take a "meta" flag like render responses, which is basically an opt-in to get our "enhanced" response format dictionary which can contain warnings.

File: api/models/graphite.go (hunk 2 above)
Comment (shanson7) [reply api/models/graphite.go]: > no one should be doing that 

Sure, but they might be. I was erring on the side of not breaking existing set ups.

> you could also just add "json"

That was just an example value. If you prefer the validation, I can add it. It just might break someone. I'll defer to you.

File: api/graphite.go (hunk 6 above)
Comment (Dieterbe) [reply api/graphite.go]: @replay shouldn't we expand the name here to include meta tags?

File: api/models/graphite.go (hunk 2 above)
Comment (Dieterbe) [reply api/models/graphite.go]: I don't think adding random undocumented and unsuppored parameters constitutes valid use of an api. if that breaks requests of people who were doing something that the spec did not specify, that's fair enough. we can always revise should anyone complain very loudly.
i just don't see why we should accommodate parameters that didn't exist before. so yeah let's add the validation.

**Comment by shanson7** [comment]
> perhaps it should also take a "meta" flag like render responses, which is basically an opt-in to get our "enhanced" response format dictionary which can contain warnings.

It seems like that would be confusing, since for `format=lastts-json` we could always make `meta=true`. Perhaps the default should be `format=graphite-json` or something and that would be the compatibility format, then `series-json` would be a dictionary format that allows `warnings` element.

**Comment by Dieterbe** [comment]
So, we can do:

1) my suggestion:
* lastts-json (default meta=true, users can provide meta=false to remove warnings. or perhaps we don't even allow setting meta=false for this format)
* series-json (default meta=false, users can opt-in)
* "" means series-json

2) your suggestion
no meta flag
* lastts-json: lastts format with warnings
* graphite-json, "", legacy format without warnings
* series-json: like graphite-json, but includes warnings

I think consistency across our different endpoints is important. It'd be really nice and clean if all endpoints had a "meta" flag which meant the same everywhere (enable metadata and richer response). Though it's annoying to have to opt-in to the better experience (if lastts-json were to default to meta=false), it'd be nicer if enabled by default. But that precludes backwards compatibility (if series-json had meta=true). We can do a comprise, and pick a different meta default based on the format variable. Which seems perhaps a bit strange, but the tradeoff seems worth it.

Proposal 2 is simple, but is different from /render and conflates orthogonal concerns (whether to include metadata/warnings with the formatting of the actual data and its encoding).  The data formatting is already conflated with encoding, E.g. What if graphite adds another output format, should we then provide an alternative metadata-enabled format string for that too (and another one when we implement another serialization function)?
Proposal 1 is a bit more complicated (additional meta parameter), but has the benefit it can be applied consistently across all endpoints, I think. (the only inconsistency is whether it's enabled by default, it seems like a more harmless inconsistency than using a completely different metadata-selecting mechanism between different http paths)

**Comment by shanson7** [comment]
> I think consistency across our different endpoints is important. It'd be really nice and clean if all endpoints had a "meta" flag which meant the same everywhere (enable metadata and richer response)

True. I think we really have 2 questions answered in various ways by different parameters: 

1. What information should be returned?
1. How should it be encoded?

For the render API, there is a `format` that is truly just the encoding (json, msgp, pickle, etc). The `meta` controls what information should be returned. I would argue that the only reason this parameter exists is to maintain graphite compatibility. Otherwise, meta would *always* be true. 

For this endpoint, we envision that there are multiple possible sets of information that would be useful to return (this PR adds 1 of them). The format parameter is answering both questions, and the meta parameter would "enhance" the decision from the first. I'm not sure in what scenario `meta=false` makes sense *except* for compatibility.

**Comment by Dieterbe** [comment]
yes if it weren't for graphite compatibility, we'd always have the rich response and no need for a meta parameter.
Your take makes sense (you say "What information should be returned?" + enhancement through meta, whereas i pulled it out a bit more into separate concerns, saying we need to choose how to format the data, and also how to format the metadata - or rather whether to include metadata - , but it's basically the same thing).

I've stated my conclusion. What's yours?

**Comment by shanson7** [comment]
> I've stated my conclusion. What's yours?

1. `lastts-json` - ignores `meta` 
1. `series-json` - compatible return if `meta=false` else JSON object with `series` and `warnings` elements

Sound good?

**Comment by Dieterbe** [comment]
so basically it's proposal 1 with the tweak that the meta flag only applies to series-json (and "" means series-json); all other formats simply include the metadata.

sounds good

**Comment by shanson7** [comment]
Excellent, because that's what I just implemented

File: api/graphite.go [hunk 8]
```diff
@@ -1025,14 +1040,45 @@ func (s *Server) graphiteTagFindSeries(ctx *middleware.Context, request models.G
 		return
 	default:
 	}
-	seriesNames := make([]string, 0, len(series))
-	for _, serie := range series {
-		seriesNames = append(seriesNames, serie.Pattern)
+
+	var warnings []string
+	if len(series) == limit {
+		warnings = append(warnings, "Result set truncated due to limit")
```
Comment (Dieterbe) [inline api/graphite.go]: this would also trigger if the amount of series happens to match the limit. maybe clusterFindByTag should explicitly set a flag when it needs to truncate.

Comment (shanson7) [reply api/graphite.go]: Yeah, I though about that, but I figured it was a reasonable edge case.

File: api/graphite.go (hunk 6 above)
Comment (replay) [reply api/graphite.go]: I can't think of a good reason why it would be necessary to include meta tags here, except for consistency so that whoever uses the endpoint also gets to see the meta tags. 
Note that in order to include meta tags in this result the `index/find_by_tags` endpoint and the `MemoryIdx.FindByTag()` method would all have to be modified to include meta tags in the returned results, which would slow finds down because those results would need to be enriched. So if we want to enable the user to also retrieve the meta tags on this endpoint we should make this an optional feature of the `index/find_by_tags` endpoint which by default is off.

Comment (shanson7) [reply api/graphite.go]: Yeah, I would say including meta tags here would be a separate feature.

File: api/graphite.go [hunk 9]
```diff
@@ -687,7 +687,12 @@ func (s *Server) executePlan(ctx context.Context, orgId uint32, plan expr.Plan)
 				return nil, meta, err
 			}
 
-			series, err = s.clusterFindByTag(ctx, orgId, exprs, int64(r.From), maxSeriesPerReq-len(reqs))
+			remainingSeriesLimit := maxSeriesPerReq - len(reqs)
+			if remainingSeriesLimit <= 0 && maxSeriesPerReq > 0 {
+				// Use 1 to enable series count checking.
```
Comment (Dieterbe) [inline api/graphite.go]: I don't understand what's going on here. This could use some clarification

File: api/graphite.go (hunk 8 above)
Comment (Dieterbe) [reply api/graphite.go]: i can live with that

**Comment by Dieterbe** [comment]
I think I was able to simplify the wording a bit.
https://github.com/grafana/metrictank/commit/cddafc4775ba65d8c15d6425a3e6227b0d4c8f90
what do you think?

File: api/graphite.go (hunk 6 above)
Comment (replay) [reply api/graphite.go]: Actually, I just looked at the related code again, and the series returned by `clusterFindByTag` already have the meta tags if the feature is enabled. They're in the `.MetaTags` property. So if that's wanted, it would be easy to also add them.

File: api/graphite.go (hunk 9 above)
Comment (shanson7) [reply api/graphite.go]: Yeah, this was a remnant from a previous iteration. I think I can clean this up

**Comment by shanson7** [comment]
@Dieterbe - I pulled in your wording changes and cleaned up the little funkiness around limit checks
